pkg/controller/common/beat/pod.go

```diff
@@ -0,0 +1,124 @@
+// Copyright Elasticsearch B.V. and/or licensed to Elasticsearch B.V. under one
+// or more contributor license agreements. Licensed under the Elastic License;
+// you may not use this file except in compliance with the Elastic License.
+
+package beat
+
+import (
+	"fmt"
+	"hash"
+
+	corev1 "k8s.io/api/core/v1"
+
+	"github.com/elastic/cloud-on-k8s/pkg/controller/common/container"
+	"github.com/elastic/cloud-on-k8s/pkg/controller/common/defaults"
+	commonhash "github.com/elastic/cloud-on-k8s/pkg/controller/common/hash"
+	"github.com/elastic/cloud-on-k8s/pkg/controller/common/volume"
+	"github.com/elastic/cloud-on-k8s/pkg/utils/pointer"
+)
+
+const (
+	CAVolumeName = "es-certs"
+	CAMountPath  = "/mnt/elastic-internal/es-certs/"
+	CAFileName   = "ca.crt"
+
+	ConfigVolumeName = "config"
+	ConfigMountPath  = "/etc/beat.yml"
+	ConfigFileName   = "beat.yml"
+
+	DataVolumeName        = "data"
+	DataMountPathTemplate = "/var/lib/%s/%s/%s-data"
+	DataPathTemplate      = "/usr/share/%s/data"
+
+	// ConfigChecksumLabel is a label used to store a Beats config checksum.
+	ConfigChecksumLabel = "beat.k8s.elastic.co/config-checksum"
+
+	// VersionLabelName is a label used to track the version of a Beat Pod.
+	VersionLabelName = "beat.k8s.elastic.co/version"
+)
+
+func buildPodTemplate(params DriverParams, defaultImage container.Image, modifyPodFunc func(builder *defaults.PodTemplateBuilder), checksum hash.Hash) corev1.PodTemplateSpec {
+	podTemplate := params.GetPodTemplate()
+
+	// Token mounting gets defaulted to false, which prevents from detecting whether user had set it.
+	// Instead, checking that here, before the default is applied.
+	// This is required for autodiscover which is enabled by default.
+	if podTemplate.Spec.AutomountServiceAccountToken == nil {
+		t := true
+		podTemplate.Spec.AutomountServiceAccountToken = &t
+	}
+
+	builder := defaults.NewPodTemplateBuilder(podTemplate, params.Type)
+
+	// might be nil if caller wants to use the default builder without any modifications
+	if modifyPodFunc != nil {
+		modifyPodFunc(builder)
+	}
+
+	builder = builder.
+		WithTerminationGracePeriod(30).
+		WithEnv(corev1.EnvVar{
+			Name: "NODE_NAME",
+			ValueFrom: &corev1.EnvVarSource{
+				FieldRef: &corev1.ObjectFieldSelector{
+					FieldPath: "spec.nodeName",
+				},
+			}}).
+		WithResources(defaultResources).
+		WithHostNetwork().
+		WithLabels(map[string]string{
+			ConfigChecksumLabel: fmt.Sprintf("%x", checksum.Sum(nil)),
+			VersionLabelName:    params.Version}).
+		WithDockerImage(params.Image, container.ImageRepository(defaultImage, params.Version)).
+		WithArgs("-e", "-c", ConfigMountPath).
+		WithDNSPolicy(corev1.DNSClusterFirstWithHostNet).
+		WithSecurityContext(corev1.SecurityContext{
+			RunAsUser: pointer.Int64(0),
```

Comment:
This is because the Beats image runs with root user by default? May be worth calling out explicitly. It also means it won't run on Openshift?

Reply:
I took this from Beats examples, but after looking more carefully I think we can do a little better.

Beat containers run as uid 1000 by default, so for most of the mounted paths it's enough if we set security context with correct fsgroup.

The issue is with hostpath volume that we use for data folder. The directory gets created with 0755 (as documented [here](https://github.com/kubernetes/kubernetes/blob/b98d9407cfdc24c477ab8066d1fb69ea0983a2f9/pkg/apis/core/types.go#L534)) and it ignores fsgroup, so uid 1000 can't write to it. If we can run init container as root we can chown it to the correct user (just like we do with elasticsearch), but if we can't then I believe the directory has to be prepared outside of k8s. [PR](https://github.com/kubernetes/kubernetes/pull/39438) to fix this was already declined.